WalletWasabi.Backend/Controllers/ChaumianCoinJoinController.cs

```diff
@@ -182,7 +187,22 @@ public async Task<IActionResult> PostInputsAsync([FromBody, Required]InputsReque
 							return BadRequest($"Input is banned from participation for {(int)bannedElem.Value.bannedRemaining.TotalMinutes} minutes: {inputProof.Input.Index}:{inputProof.Input.TransactionId}.");
 						}
 
-						GetTxOutResponse getTxOutResponse = await RpcClient.GetTxOutAsync(inputProof.Input.TransactionId, (int)inputProof.Input.Index, includeMempool: true);
+						var txoutResponseTask = batch.GetTxOutAsync(inputProof.Input.TransactionId, (int)inputProof.Input.Index, includeMempool: true);
+						getTxOutResponses.Add((inputProof, txoutResponseTask));
+					}
+
+					// Perform all RPC request at once
+					var waiting = Task.WhenAll(getTxOutResponses.Select(x=>x.getTxOutTask));
```

Comment:
This batching logic may be dangerous. I will review carefully later.

Reply:
If the other side (NBitcoin) is threadsafe than this should not be a problem. Tested several times here:
https://github.com/MetacoSA/NBitcoin/blob/master/NBitcoin.Tests/RPCClientTests.cs
On regtest the coinjoin is working.

Reply:
I've ran the `CcjTestsAsync` test while observing the local network traffic with a `tcpdump` session and this is what it does:
 
```
POST / HTTP/1.1
Authorization: Basic NDE4MzM0ZDc1ZWE5NGJhNzRlZDdiNjVhNzRkYjU1YTBhNGMyN2E5Zjo0MTgzMzRkNzVlYTk0YmE3NGVkN2I2NWE3NGRiNTVhMGE0YzI3YTlm
Connection: close
Content-Type: application/json-rpc
Content-Length: 911
Host: 127.0.0.1:11775

[{"jsonrpc":"1.0","id":1,"method":"gettxout","params":["41c3700ea541c38ec4818a6d0270d0fbb932874374b04e015567bfe62de54257",0,true]},
{"jsonrpc":"1.0","id":1,"method":"gettxout","params":["1674e77274e1e8ef3cf9a5ab1fbac0302e07f8eaf31f54e83dae5be2f485610a",0,true]},
{"jsonrpc":"1.0","id":1,"method":"gettxout","params":["99fc02f245690cee8347449e8de445d9efac09705d33cb28c24fc02de2c8e81c",0,true]},
{"jsonrpc":"1.0","id":1,"method":"gettxout","params":["ee91b463c3ba37d44e0f275f0bbe1b1fcc7002539b9c2e7b87328c733dfb144e",1,true]},
{"jsonrpc":"1.0","id":1,"method":"gettxout","params":["848a7e2870a6db7a6515cabf3d94e7f3dab324377fe4c151cc312fadae1ac6c0",0,true]},
{"jsonrpc":"1.0","id":1,"method":"gettxout","params":["9d9594d0e3bebb796f4f008e60075f40e4fddef4a59a23c7868280c5a80ea347",0,true]},
{"jsonrpc":"1.0","id":1,"method":"gettxout","params":["ab9617f6bf8c627d4be8d7bb24966235944d13c16c706adfd52e0b62fce4954f",0,true]}]
```

```
HTTP/1.1 200 OK
Content-Type: application/json
Date: Wed, 02 Jan 2019 17:17:42 GMT
Content-Length: 2718
Connection: close

[{"result":"bestblock":"5680dda2854319e9aa2392eac3b6937854f9cc308e1f7f5a0d1d7a5299d53b72","confirmations":9,"value":0.01000000,"scriptPubKey":{"asm":"0 6e62a31570f03f8202829011a85a6969ec147e2d","hex":"00146e62a31570f03f8202829011a85a6969ec147e2d","reqSigs":1,"type":"witness_v0_keyhash","addresses":["bcrt1qde32x9ts7qlcyq5zjqg6sknfd8kpgl3dje3yg8"]},"coinbase":false},"error":null,"id":1},
{"result":{"bestblock":"5680dda2854319e9aa2392eac3b6937854f9cc308e1f7f5a0d1d7a5299d53b72","confirmations":8,"value":0.01000000,"scriptPubKey":{"asm":"0 ecd199ae5f685966fb591b8a085e957960e8b14f","hex":"0014ecd199ae5f685966fb591b8a085e957960e8b14f","reqSigs":1,"type":"witness_v0_keyhash","addresses":["bcrt1qangentjldpvkd76erw9qsh5409sw3v20k98ycw"]},"coinbase":false},"error":null,"id":1}
{"result":{"bestblock":"5680dda2854319e9aa2392eac3b6937854f9cc308e1f7f5a0d1d7a5299d53b72","confirmations":7,"value":0.01000000,"scriptPubKey":{"asm":"0 481b2b4301b5329cc2aa322f2097cd808c7656d1","hex":"0014481b2b4301b5329cc2aa322f2097cd808c7656d1","reqSigs":1,"type":"witness_v0_keyhash","addresses":["bcrt1qfqdjkscpk5efes42xghjp97dszx8v4k3daslqp"]},"coinbase":false},"error":null,"id":1},
{"result":{"bestblock":"5680dda2854319e9aa2392eac3b6937854f9cc308e1f7f5a0d1d7a5299d53b72","confirmations":6,"value":0.01000000,"scriptPubKey":{"asm":"0 1017cb94fb69f565772db825bfb09e628365dc19","hex":"00141017cb94fb69f565772db825bfb09e628365dc19","reqSigs":1,"type":"witness_v0_keyhash","addresses":["bcrt1qzqtuh98md86k2aedhqjmlvy7v2pkthqecummgr"]},"coinbase":false},"error":null,"id":1},
{"result":{"bestblock":"5680dda2854319e9aa2392eac3b6937854f9cc308e1f7f5a0d1d7a5299d53b72","confirmations":5,"value":0.01000000,"scriptPubKey":{"asm":"0 f0a1ac427f5e6291540542b125a826d1d6b65c7c","hex":"0014f0a1ac427f5e6291540542b125a826d1d6b65c7c","reqSigs":1,"type":"witness_v0_keyhash","addresses":["bcrt1q7zs6csnlte3fz4q9g2cjt2px68ttvhrufc29hm"]},"coinbase":false},"error":null,"id":1},
{"result":{"bestblock":"5680dda2854319e9aa2392eac3b6937854f9cc308e1f7f5a0d1d7a5299d53b72","confirmations":4,"value":0.01000000,"scriptPubKey":{"asm":"0 49ee65f68554fdce3621ad5d8d623a055528b911","hex":"001449ee65f68554fdce3621ad5d8d623a055528b911","reqSigs":1,"type":"witness_v0_keyhash","addresses":["bcrt1qf8hxta592n7uud3p44wc6c36q42j3wg32enhyk"]},"coinbase":false},"error":null,"id":1},
{"result":{"bestblock":"5680dda2854319e9aa2392eac3b6937854f9cc308e1f7f5a0d1d7a5299d53b72","confirmations":3,"value":0.01000000,"scriptPubKey":{"asm":"0 34fd69a5b0755f1669d179e1dc59464312a5cd34","hex":"001434fd69a5b0755f1669d179e1dc59464312a5cd34","reqSigs":1,"type":"witness_v0_keyhash","addresses":["bcrt1qxn7knfdsw403v6w308sack2xgvf2tnf59nr63n"]},"coinbase":false},"error":null,"id":1}]
```